### PR TITLE
integrate aligner for phoneme overampling

### PR DESCRIPTION
**Pull request by manmay-nakhashi**
quick aligner integration from naturalspech2 repo

### CODE REVIEW

**Comment by manmay-nakhashi** [comment]
@lucidrains done

File: voicebox_pytorch/voicebox_pytorch.py [hunk 1]
```diff
@@ -389,8 +423,11 @@ def forward(
         num = reduce(loss, 'b n -> b', 'sum')
         den = mask.sum(dim = -1).clamp(min = 1e-5)
         loss = num / den
-
-        return loss.mean()
+        loss.mean()
```
Comment (lucidrains) [inline voicebox_pytorch/voicebox_pytorch.py:426]: missing assignment

File: voicebox_pytorch/voicebox_pytorch.py [hunk 2]
```diff
@@ -389,8 +423,11 @@ def forward(
...
+        
+        #aligner loss
+        loss += self.align_loss(alignment_logprob, phoneme_len, mel_len)
```
Comment (lucidrains) [inline voicebox_pytorch/voicebox_pytorch.py:429]: safer to use `loss = loss + `

File: voicebox_pytorch/voicebox_pytorch.py [hunk 3]
```diff
@@ -328,12 +331,36 @@ def forward_with_cond_scale(
...
+            - alignment_hard: `[B, T]`
+            - alignment_soft: `[B, T_x, T_y]`
+            - alignment_logprob: `[B, 1, T_y, T_x]`
+            - alignment_mas: `[B, T_x, T_y]`
+        """
+        attn_mask = rearrange(x_mask, 'b 1 t -> b 1 t 1') * rearrange(y_mask, 'b 1 t -> b 1 1 t')
+        alignment_soft, alignment_logprob = self.aligner(rearrange(y, 'b t c -> b c t'), x, x_mask)
+        assert not torch.isnan(alignment_soft).any()
+        alignment_mas = maximum_path(
+            rearrange(alignment_soft, 'b 1 t1 t2 -> b t2 t1').contiguous(), attn_mask.squeeze(1).contiguous()
```
Comment (lucidrains) [inline voicebox_pytorch/voicebox_pytorch.py:353]: could you replace the squeeze with einops too? i know, i'm an einops fanboy

File: voicebox_pytorch/voicebox_pytorch.py [hunk 4]
```diff
@@ -295,7 +295,10 @@ def __init__(
             kernel_size = conv_pos_embed_kernel_size,
             groups = conv_pos_embed_groups
         )
-
+        # if we are using mel spec with 80 channels, we need to set attn_channels to 80
+        # dim_in assuming we have spec with 80 channels
+        self.aligner = Aligner(dim_in = 80, dim_hidden = dim_phoneme_emb, attn_channels = 80)
```
Comment (lucidrains) [inline voicebox_pytorch/voicebox_pytorch.py:300]: we could probably have the `dim_in` and `attn_channels` as hyperparameters, perhaps passed in as `aligner_kwargs: dict = dict(dim_in = 80, attn_channels = 80)` as default

**Comment by lucidrains** [comment]
@manmay-nakhashi looking better! left a few comments!

**Comment by lucidrains** [comment]
@manmay-nakhashi thank you!
